### PR TITLE
[FLOSS H3] File name checker for CI

### LINKED ISSUE #1615: [FLOSS H1] build server check: only allow a-z, 0-9 and _ for file names
We have inconsistent use of characters in file names, which always requires one to think which kind of separator is used in specific cases.

```
~e> find . -name '*-*' | wc -l
386
~e> find . -name '*_*' | wc -l
491
~e> find . -name '* *' | wc -l
0
```

As short time goal we should not introduce new variants (such as spaces, see #1613), as 1.0 goal I propose to unify it to a single variant.

### PR DESCRIPTION
**Pull request by muskater**
This will fix #1615. I have created a shell script that check the file names of the whole repository. It is integrated in the Cirrus CI. I tried to make  comments and explain my thought process in the shell scrip. Suggestions on how to implement it more elegantly or robust are very welcome! 

## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [s] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: scripts/filename-checker [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
Comment (markus2330) [inline scripts/filename-checker:1]: The file should be in scripts/dev. Then you probably need another script in scripts/build/ which runs the filename-checker for every file, similar to what the scripts/dev/reformat- are doing.

File: scripts/filename-checker [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+if [ $# -eq 0 ]; then
+	echo "Usage: filename-checker <path/to/file>"
+	exit
+fi
+
+SCRIPT_DIR="$(dirname "$0")"
+
+ALLOWLIST=$(mktemp)
+grep -Ev '^(#|[[:space:]]*$)' "$SCRIPT_DIR/../tests/filename.whitelist" > "$ALLOWLIST"
```
Comment (markus2330) [inline scripts/filename-checker:16]: The script needs to fail if the file couldn't be found. The check only checks for spaces, it should be more strict.

**Review by markus2330 — COMMENTED** [review]
Some comments.

**Comment by markus2330** [comment]
Is there any specific question?

**Comment by muskater** [comment]
## Some Questions

I am currently working on implementing a way for the CI system to check if a newly added file is compliant with the libelektra file naming conventions.  I would like to take a similar approach to the link checker.
Furthermore, I have got some questions on how I should tackle this task:

Is using a shell script like the link-checker a sensible idea?

I want to have an allow-list of file-names that cannot comply to rules. Does this make sense?

## Some Implementation Ideas

My idea thus far is to have a script or the CI run the following command in the cloned repository 

`find . | grep -v '^[-_/.a-zA-Z0-9]*$'`

Then I would compare the output with the allow-list. If they match, everything is fine; else the check would fail.

Would I actually need a full on shell script for that, or does it make more sense to implement in `cirrus.yml`?

**Comment by muskater** [comment]
> Is there any specific question?

Sorry it took me some time to formulate my questions :)

**Comment by markus2330** [comment]
> Is using a shell script like the link-checker a sensible idea?

Yes. The shell script can easily be invoked from Jenkins. scripts/jenkins/Jenkinsfile e.g. like scripts/build/run_check_release_notes

> I want to have an allow-list of file-names that cannot comply to rules. Does this make sense?

Yes, this part already looks good.

> Then I would compare the output with the allow-list. If they match, everything is fine; else the check would fail.

I think its easier to directly let the script fail, without writing to a temporary file.

**Comment by markus2330** [comment]
Btw. you do not need the script to check for a single file. It is enough to have a script that checks for all files (with a fine statement similar to already written excluding file names from a whitelist).

**Comment by muskater** [comment]
@markus2330 Thank you very much for the fast response!

**Comment by markus2330** [comment]
Please ask if there is any further question. Make sure to document what the scripts should do, then it is easier to help you writing the scripts.

**Comment by markus2330** [comment]
And another tip: start with the most easy script and add features (like whitelist of filenames) later.

**Comment by muskater** [comment]
Quick Update: The main functionality of the script should work now. Some more testing and new test cases are still needed though.
I tried to make more comments and explain my thought process. Suggestions on how to implement the script more elegantly or robust are very welcome! I am fairly new to shell scripting.

Because I do not want to break the CI I did not push the changes that I have planed to do. My Idea is to include the script in the cirrus CI similar to the link checker. Here is what I got so far:

```
task:
  matrix:

    - name: 📄 Check
      container:
        cpu: 2
        memory: 4G
        dockerfile: scripts/docker/cirrus/arch/Dockerfile

  clone_script: |
    if [ -z "$CIRRUS_PR" ]; then
      git clone --branch=$CIRRUS_BRANCH https://x-access-token:${CIRRUS_REPO_CLONE_TOKEN}@github.com/${CIRRUS_REPO_FULL_NAME}.git \
                $CIRRUS_WORKING_DIR
      git reset --hard $CIRRUS_CHANGE_IN_REPO
    else
      git clone https://x-access-token:${CIRRUS_REPO_CLONE_TOKEN}@github.com/${CIRRUS_REPO_FULL_NAME}.git $CIRRUS_WORKING_DIR
      git fetch origin pull/$CIRRUS_PR/head:pull/$CIRRUS_PR
      git config user.email "nobody@example.com"
      git config user.name "nobody"
      git merge --no-commit --no-ff $CIRRUS_CHANGE_IN_REPO
    fi

  script:
    - *generate
    - scripts/dev/filename-checker > bad_file_names.txt

  tests_script:
    - |
      if test -s broken_links.txt; then
        printf >&2 '—————————————\n'
        cat >&2 bad_file_names.txt
        printf >&2 '—————————————\n'
        false
      fi

```

**Comment by muskater** [comment]
Is there a simple way how I could test the CI changes locally?

**Comment by markus2330** [comment]
As CI does nothing as running the script in a Docker container, you can simply test by running the script in a Docker container.

For testing the Jenkinsfile, you need to push to this branch and see what Jenkins is doing.

**Comment by markus2330** [comment]
> Because I do not want to break the CI I did not push the changes that I have planed to do.

The PR is isolated, so you do not need to worry about that.

**Comment by lawli3t** [comment]
Very good idea! Moving this to Jenkins seems not so bad, because being dependent on the free tiers of CIs can be a bit iffy at times. Adding this to Jenkins as well would be huge I think.

**Comment by muskater** [comment]
Quick update: I have added the file name checker to the Jenkins file. 
(I will definitely squash my changes once this PR is ready to merge btw.)

File: scripts/dev/filename-checker [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# find any file that is non-compliant
+# save them to the variable FORBIDDEN_NAMES
+FORBIDDEN_NAMES=$(find . | grep -v '^[-_/.a-zA-Z0-9]*$')
+
+# All file names that offend the rule are saved in band_names
+BANNED_NAMES=$(echo "$FORBIDDEN_NAMES" | while read -r test_line; do
+	# Set the delimiter to newline to compare every line
+	# import the variable test_line as l
+	# compare line by line the allow-list with the read line
+	# if a match is found write that to check
```
Comment (markus2330) [inline scripts/dev/filename-checker:33]: I do not understand the sentence. Was this comment updated? Probably it should be removed, as nothing is written?

File: scripts/dev/filename-checker [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	CHEK=$(echo "$ALLOW_LIST" | awk -F "\n" -v l="$test_line" '$1 == l { print $1 }')
```
Comment (markus2330) [inline scripts/dev/filename-checker:34]:
```suggestion
	CHECK=$(echo "$ALLOW_LIST" | awk -F "\n" -v l="$test_line" '$1 == l { print $1 }')
```

File: scripts/dev/filename-checker [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	# When check is empty it means that that no entry in the white list was found
+	if [ -z "$CHEK" ]; then
```
Comment (markus2330) [inline scripts/dev/filename-checker:36]:
```suggestion
	if [ -z "$CHECK" ]; then
```

File: scripts/dev/filename-checker [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+		echo "$test_line"
+	fi
+done)
+
+if [ -n "$BANNED_NAMES" ]; then
+	echo >&2 "These names do not comply with the naming convention: [/.a-zA-Z0-9_-]*"
+	echo >&2 "Please change them accordingly or add them to the allow-list"
```
Comment (markus2330) [inline scripts/dev/filename-checker:43]:
```suggestion
	echo >&2 "Please change them accordingly or add them to tests/filename.whitelist"
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Looks already much better!

**Comment by muskater** [comment]
Thank you for the great suggestions. I tried to make the comments more consistent and include your changes.

**Comment by muskater** [comment]
Should I add a line to the pull request template that mentions the file name checker? For example like:

- [ ] make sure all newly added files are named according to `[/.a-zA-Z0-9_-]*` . Exceptions can be added to `tests/filename.whitelist`

**Comment by markus2330** [comment]
> Should I add a line to the pull request template that mentions the file name checker?

Good idea but I think this error is too specific and rare to let people manually check for it. That the CI checks for it is enough. :fireworks:

**Comment by muskater** [comment]
There where some merge conflicts in the Jenkins file, but I was able to solve them manually. I hope I did not break any thing.

**Pull request merged**